### ISSUE TITLE
Extends OrientDB SQL with new functions

### ISSUE THREAD

**Issue by lvca** [orientechnologies/orientdb#2573]
## FOR functions

Like Java for cycle. Syntax:

```
for( <initial-expression>; <expression to check at every cycle>; <expression to execute at every cycle> )
```
## Math functions

A good idea could be bind all the math functions in one shot by wrapping the Java Math class. Syntax: 

```
math.<math-function(<args>*)
```

Example:

  select math.sin( x ) from points

(`split()` already supported in 2.2)

**Comment by fabriziofortino** [comment]
@lvca I have implemented a new OSQLFunction that uses reflection to easily bind any static method (with overload support). It is extremely easy to bind all the static methods of a given class. I have added the support for the java.lang.Math methods. The only problem is the prefix: I cannot used dots in the functions' names (eg: math.random). OSQLFilterItemAbstract (line 59) splits the string removing the part before the dot. I have currently removed the prefix but some names clash with the existing functions (eg: max, min). Any idea on how to sort it out?

You can see the changes here:
https://github.com/fabriziofortino/orientdb/commit/71f37b34717a3f5544c1cfd1fca9432644c340f9

**Comment by lvca** [comment]
We could introduce the concept of package in function (we should have some issues about it). What about: `Math::max()` ?

**Comment by fabriziofortino** [comment]
The introduction of packages sounds good to me. I guess we should modify the parser to support it. I have no experience with javacc though. Any luck to add this feature soon? 

With the current implementation, adding the prefix you proposed I get the following exception

```
com.orientechnologies.orient.core.sql.parser.ParseException: Encountered " ":" ": "" at line 1, column 12.
Was expecting one of:
    <EOF> 
    <TIMEOUT> ...
    <AS> ...
    <FETCHPLAN> ...
    <LOCK> ...
    <LET> ...
    <NOCACHE> ...
    <PARALLEL> ...
    <UNWIND> ...
    ";" ...
    "," ...
    <AS> ...
    "," ...
    
	at com.orientechnologies.orient.core.sql.parser.OrientSql.generateParseException(OrientSql.java:11559)
	at com.orientechnologies.orient.core.sql.parser.OrientSql.jj_consume_token(OrientSql.java:11432)
	at com.orientechnologies.orient.core.sql.parser.OrientSql.parse(OrientSql.java:88)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLAbstract.preParse(OCommandExecutorSQLAbstract.java:234)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:201)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLSelect.parse(OCommandExecutorSQLSelect.java:1)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLResultsetAbstract.parse(OCommandExecutorSQLResultsetAbstract.java:1)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:74)
	at com.orientechnologies.orient.core.sql.OCommandExecutorSQLDelegate.parse(OCommandExecutorSQLDelegate.java:40)
	at com.orientechnologies.orient.core.storage.impl.local.OAbstractPaginatedStorage.command(OAbstractPaginatedStorage.java:1181)
```

**Comment by lvca** [comment]
@luigidellaquila please could you work with @fabriziofortino on this?

**Comment by luigidellaquila** [comment]
Hi @fabriziofortino

dot notation for packages is not an alternative, because it's already used for path navigation, eg. math.random() means invoke "random" method on "math" attribute.

I like very much the package::function syntax, I can implement it very quickly in the new parser, but I have to check what is the impact on the rest of the stack

Luigi

**Comment by fabriziofortino** [comment]
@luigidellaquila package::function syntax sounds good to me. Let me know when you implement it so I can adjust my changes

**Comment by fabriziofortino** [comment]
@luigidellaquila any news on this?

**Comment by luigidellaquila** [comment]
Hi Fabrizio,

not actually, sorry. It's in the queue, but right now we are focused on 2.1
final release, so I'm afraid you have to wait a few days ;-)

Luigi


2015-05-21 10:28 GMT+02:00 Fabrizio Fortino <notifications@github.com>:

**Comment by fabriziofortino** [comment]
@lvca @luigidellaquila any luck to have the package::function syntax in 2.2 so I can extend the SQL with java.lang.Math functions?

**Comment by luigidellaquila** [comment]
hi @fabriziofortino 

yes, I guess it's time to implement this feature.
My only concern is about non-standard syntax: writing a query like

```sql
select math::max(age) from Person
```
is not a problem, I parse the query and I have full control on this.
The problem arises when you want to use this function inside another Javascript function, in this case this syntax does not work. Moreover, the expected behavior would be that functions in the same package can call each other without package prefix, but it brings other problems, because we should refactor a lot current js command executors.

No big deal of course, we could just decide to drop all these features and let package functions available only through SQL.

**Comment by fabriziofortino** [comment]
@luigidellaquila I reckon that having the package feature available from SQL only is a good compromise

**Comment by lvca** [comment]
I like the concept of package visibility, but we should ind a way to call cross package function from JS, because :: is not supported. What's the JS mode to do a package?

**Comment by luigidellaquila** [comment]
there is no package concept in JS, you just can have classes that contain functions.

A solution could be
```sql
select math().max(foo) from P
```
where ```math()``` returns an object with mathematical functions as methods

**Comment by fabriziofortino** [comment]
Probably the JS ways that resemble package concept are namespaces and modules (http://www.kenneth-truyers.net/2013/04/27/javascript-namespaces-and-modules/)

But in both cases to access a function inside a namespace/module we have to use the dot notation

```sql
select math.max(x1, x2) from myClass
```

that clashes with the path navigation syntax.

**Comment by fabriziofortino** [comment]
@luigidellaquila any news on this? Apart from the math functions I have created other functions to perform matching operations that I'd like to propose after we figure out how to solve the package / prefix problem.

**Comment by luigidellaquila** [comment]
Hi @fabriziofortino 

we discussed a lot on this topic and our conclusion was that for now we prefer to avoid additional concepts like namespaces, so in the medium term we will go with simple prefixes.

We are doing the same with Lucene, with "ST_" functions.

Thanks

Luigi

**Comment by lvca** [comment]
It would be nice creating a piece of code that auto-register all `Math.<method>()` automatically with `math_` prefix. Example: `math_abs`.

**Comment by fabriziofortino** [comment]
@luigidellaquila sounds good to me.

@lvca I have already done something similar in the first PR.

I will submit another PR for the `java.lang.Math.*` function using the `math_` prefix.

**Comment by fabriziofortino** [comment]
@luigidellaquila one more question. Since the Lucene functions use uppercase prefix (`ST_`) should I use the same convention for the math ones (`MATH_` instead of `math_`)?

**Comment by lvca** [comment]
Function names should be case insensitive.

**Comment by mmacfadden** [comment]
This is something I would be willing to help with.

**Comment by andreafalzetti** [comment]
Hi guys, any update on this?

**Comment by fabriziofortino** [comment]
@andreafalzetti I did some work on this to bind all the Java Math methods into sql functions (commit is here https://github.com/fabriziofortino/orientdb/commit/71f37b34717a3f5544c1cfd1fca9432644c340f9).

I didn't follow up on this though cause I was busy with other stuff. I'd say I need to adapt the code to the latest changes since the code has almost 2 years now. I will be happy to work on this in the next days.

@luigidellaquila What do you think? Should these new functions be implemented in 2.2.x or 3.x ?

**Comment by andreafalzetti** [comment]
It would be great to see this implemented @fabriziofortino
